cfgov/ask_cfpb/models/search/documents.py

```diff
@@ -0,0 +1,508 @@
+        answer_text = self.answer_content_text()
+        full_text = f"{short_answer}\n\n{answer_text}\n\n{self.question}"
+        return full_text
+
+    def __str__(self):
+        if self.answer_base:
+            return f"{self.answer_base.id}: {self.title}"
+        else:
+            return self.title
+
+    @property
+    def clean_search_tags(self):
+        return [
+            tag.strip()
+            for tag in self.search_tags.split(",")
+        ]
+
+    @property
+    def status_string(self):
+        if self.redirect_to_page:
+            if not self.live:
+                return ("redirected but not live")
+            else:
+                return ("redirected")
+        else:
+            return super(AnswerPage, self).status_string
+
+    # Returns an image for the page's meta Open Graph tag
+    @property
+    def meta_image(self):
+        if self.social_sharing_image:
+            return self.social_sharing_image
+
+        if not self.category.exists():
+            return None
+
+        return self.category.first().category_image
+
+    # Overrides the default of page.id for comparing against split testing
+    # clusters. See: core.feature_flags.in_split_testing_cluster
+    @property
+    def split_test_id(self):
+        return self.answer_base.id
+
+
+@registry.register_document
+class AnswerPageSearchDocument(Document):
+
+    autocomplete = fields.TextField(analyzer=label_autocomplete)
+    portal_topics = fields.KeywordField()
+    portal_categories = fields.TextField()
+    text = fields.TextField(attr="text", analyzer=synonym_analyzer)
+    url = fields.TextField()
+    preview = fields.TextField(attr="answer_content_data")
+
+    def get_queryset(self):
+        query_set = super().get_queryset()
+        return query_set.filter(live=True, redirect_to_page=None)
+
+    def prepare_autocomplete(self, instance):
+        return instance.question
+
+    def prepare_portal_categories(self, instance):
+        return [topic.heading for topic in instance.portal_category.all()]
+
+    def prepare_portal_topics(self, instance):
+        return [topic.heading for topic in instance.portal_topic.all()]
+
+    def prepare_search_tags(self, instance):
+        return instance.clean_search_tags
+
+    def prepare_url(self, instance):
+        return instance.url
+
+    class Index:
+        name = 'ask-cfpb'
+        settings = {'number_of_shards': 1,
+                    'number_of_replicas': 0}
+
+    class Django:
+        model = AnswerPage
+
+        fields = [
+            'search_tags',
+            'language',
+        ]
+
+
+def get_suggestion_for_search(search_term):
+    s = AnswerPageSearchDocument.search().suggest(
+        'text_suggestion', search_term, term={'field': 'text'})
+    response = s.execute()
+    try:
+        suggested_term = response.suggest.text_suggestion[0].options[0].text
+        return suggested_term
+    except IndexError:
+        return search_term
+
+
+class AnswerPageSearch:
```

Comment:
I'm not sure that the `models` package is the right place for this to live — it's not a data model. 

I also agree with @alexm118, it would be nice to generalize this class and put it in `cfgov.search`. It looks like the only thing we'd have to do is maybe add a `document_class` argument to `__init__` (or use `base_query` and require it?) and use that instead of using `AnswerPageSearchDocument` directly. That way any of our searches could pass a `Document` subclass and get autocomplete, search, and suggestions without reimplementing any of this.

Reply:
Agreed, I think we can make this a more generalized search class, as I believe the logic for implementing the autocomplete, search, and suggest features will be largely the same with the only outlier being the field within the Document we act upon. I'm not sure fully how that will look but I think giving it a shot and evaluating its utility over time could prove worthwhile.

Reply:
Moved AnswerPageSearch class from ask_cfpb models into cfgov.search

Reply:
@willbarton Please expand on how to implement your suggestion as to which specific class you would like to add a `document_class` argument to `__init__`. Currently the search only supports ask_cfpb. I like the idea of making the search generic and support passing a `Document` subclass.